### ISSUE TITLE
Need bulk operations within resource matching module

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#487]
For more complex queuing policies like conservative backfill, each schedule loop needs to issue many `match-allocate` or `match-allocate-orelse-reserve` RPCs. To avoid the excessive RPC latency incurred by many RPC issuance, we need to extend the individual job-based RPC to support "bulk operation". 

For example, a bulk RPC within `resource` should support `match-allocate` or `cancel` on a list of jobs in addition to a single job.

**Comment by dongahn** [comment]
During our last Thu meeting, I mentioned that the current bottleneck for `qmanager`/`resource` seemed like RPC costs. @garlick and @grondo suggested that we should be able to get nearly equivalent performance improvement by using asynchronous RPC. 

This is a great suggestion. One caveat, though: there are queuing policies that require strict ordering on an issuance or type of a match request. It depends on the result of its previous match request. As a simple case, a match request received by `resource` must not start if its previous match request could not be granted if the queuing policy is FCFS.

I can think of various schemes to get around this. But a good generalized scheme requires more design thoughts. BTW, even if we implement a bulk match, we will have to deal with this ordering issues any way.

**Comment by dongahn** [comment]
Change the title to reflect the direction better.

**Comment by garlick** [comment]
Sorry for the over-simplification.  It sounds like @tpatki's suggestion of a batch protocol was spot on.

**Comment by SteVwonder** [comment]
> BTW, even if we implement a bulk match, we will have to deal with this ordering issues any way.

I agree, I don't think the bulk protocol saves us any effort, but it may get us performance.  If we encode extra information into the protocol (e.g., how many reservations still need to be made before the scheduler can do backfilling), we will probably get a performance boost.  On the other hand, it will require an entirely separate callback in the resource matching service and breakdown the really nice separation of concerns that exists between `qmanager` and `resource`.

I really like @garlick's async proposal as a first attempt at performance improvement.  It keeps the separation of concerns and shouldn't require any extra code in `resource`.

> there are queuing policies that require strict ordering on an issuance or type of a match request.
> As a simple case, a match request received by resource must not start if its previous match request could not be granted if the queuing policy is FCFS.

Also a really good point.  AFAICT for FCFS the async solution doesn't buy us much.  Luckily, I think the common use-case for `qmanager` and `resource` will be some form of backfilling.  The best case for the async solution seems to be straightforward backfilling with no reservations.  The outcome of any given allocation does not affect later ones.  In this case, `qmanager` can send an unbounded number of async requests to `resource` without affecting correctness.

I think backfilling with reservations can also benefit from the async solution.  This just requires a two-phase approach.  The first phase is getting through the reservations, and the second phase is essentially straight backfilling.

**Comment by dongahn** [comment]
Good feedback @garlick and @SteVwonder.

> Also a really good point. AFAICT for FCFS the async solution doesn't buy us much. 

I actually think asyn RPC can benefit FCFS as well by pipling those jobs that can be scheduled back to back.

Like you said, I think the key is to augment resource's match request callback in a way not to break the current separation of concerns.

At first glance, it feels like adding the result of the previous match into the semantics of match in combination with having one more type or errno beyond `EBUSY` and `ENODEV`.

When I get to this, I will create a table with new semantics that can take into account the result of the previous match...

And yes backfill is will be the most straightforward beneficiary of this. 

I just need to devise the protocol so that this equally benefits all of the general queuing and backfilling policies.

**Comment by dongahn** [comment]
Label duplicate as Issue #577
